dbms/src/main/java/org/polypheny/db/processing/AbstractQueryProcessor.java

```diff
@@ -283,98 +327,159 @@ public PolyphenyDbSignature prepareQuery( RelRoot logicalRoot, RelDataType param
         if ( isAnalyze ) {
             statement.getProcessingDuration().start( "Implementation Caching" );
         }
-        if ( RuntimeConfig.IMPLEMENTATION_CACHING.getBoolean() && statement.getTransaction().getUseCache() && (!routedRoot.kind.belongsTo( SqlKind.DML ) || RuntimeConfig.IMPLEMENTATION_CACHING_DML.getBoolean() || statement.getDataContext().getParameterValues().size() > 0) ) {
-            PreparedResult preparedResult = ImplementationCache.INSTANCE.getIfPresent( parameterizedRoot.rel );
-            if ( preparedResult != null ) {
-                PolyphenyDbSignature signature = createSignature( preparedResult, routedRoot, resultConvention, executionTimeMonitor );
-                if ( isAnalyze ) {
-                    statement.getProcessingDuration().stop( "Implementation Caching" );
-                }
 
-                //TODO @Cedric this produces an error causing several checks to fail. Please investigate
-                //needed for row results
-
-                //final Enumerable enumerable = signature.enumerable( statement.getDataContext() );
-                //Iterator<Object> iterator = enumerable.iterator();
-
-                if ( statement.getTransaction().getMonitoringData() != null ) {
-                    StatementEvent eventData = statement.getTransaction().getMonitoringData();
-                    eventData.setMonitoringType( parameterizedRoot.kind.sql );
-                    eventData.setDescription( "Test description: " + signature.statementType.toString() );
-                    eventData.setRouted( logicalRoot );
-                    eventData.setFieldNames( ImmutableList.copyOf( signature.rowType.getFieldNames() ) );
-                    //eventData.setRows( MetaImpl.collect( signature.cursorFactory, iterator, new ArrayList<>() ) );
-                    eventData.setAnalyze( isAnalyze );
-                    eventData.setSubQuery( isSubquery );
-                    //eventData.setDurations( statement.getProcessingDuration().asJson() );
+        for ( int i = 0; i < proposedRoutingPlans.size(); i++ ) {
+            RelRoot routedRoot = proposedRoutingPlans.get( i ).getRoutedRoot();
+            if ( this.isImplementationCachingActive( statement, routedRoot ) ) {
+                RelRoot parameterizedRoot = parameterizedRootList.get( i );
+                PreparedResult preparedResult = ImplementationCache.INSTANCE.getIfPresent( parameterizedRoot.rel );
+                RelNode optimalNode = QueryPlanCache.INSTANCE.getIfPresent( parameterizedRootList.get( i ).rel );
+                if ( preparedResult != null ) {
+                    PolyphenyDbSignature<?> signature = createSignature(
+                            preparedResult,
+                            parameterizedRoot.kind,
+                            optimalNode,
+                            parameterizedRoot.validatedRowType,
+                            resultConvention,
+                            executionTimeMonitor );
```

Comment:
This means, we create a signature for every candidate plan. @cedricmendelin  Is there a particular reason for doing this? If not, we should create the signature at the end of the prepareQuery() method.

Reply:
No, I do not know if there is a particular reason. Maybe not all information are available later on? But if so, I do not see a reason why it should not be able to generate at the end.

Reply:
In this case, we should do this. 

@datomo Have you already looked into implementing a new wrapper type for implemented query plans? If so, we might leave this as it is for this PR and do this refactoring together with the introduction of the new type. What do you think?

Reply:
No, not yet, I am still consumed by the renaming. But I will definitly change the signature in my PR, so it should be fine if left for this PR.

Reply:
Okay, perfect. Then we'll leave it as it is for this PR.